### PR TITLE
kfctl functional description and changes required

### PR DESCRIPTION
**Pull request by kkasravi**
Related to #1419 

kfctl functional description in README.md

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kubeflow/2091)

### CODE REVIEW

File: kfctl/README.md [hunk 1]
```diff
@@ -0,0 +1,140 @@
+# kfctl golang client
+
+## Overview
+
+The kfctl golang client will provide a CLI similar to kfctl.sh but will 
+address the following deficits in the current tool
+
+1. No longer require a local install of ks (ksonnet)
```
Comment (jlewi) [inline kfctl/README.md]: Can you explain this goal? Do you intend to duplicate the functionality of ks? 
The intent with kfctl.sh was that if you wanted to customize individual components you would have to go into the ksonnet app directory?

File: kfctl/README.md [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+2. No longer require knowledge of the ks CLI 
+3. No longer save a ksonnet application to the user's filesystem
```
Comment (jlewi) [inline kfctl/README.md]: Why don't you want to save the ksonnet application to the user's filesystem?

File: kfctl/README.md [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+4. Work in a similar way as the kubeflow UI [gcp-click-to-deploy](https://github.com/kubeflow/kubeflow/tree/master/components/gcp-click-to-deploy):
+  - Execution of subcommands `init`, `generate`, `apply` and `delete` are done by the bootstrapper.
```
Comment (jlewi) [inline kfctl/README.md]: Do you mean they aren't running locally? Why aren't they running locally?

File: kfctl/README.md [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+cd myapp
+kfctl.sh generate all
+kfctl.sh apply all
+```
+
+kfctl (golang) usage is similar:
+
+```sh
+kfctl init myapp 
+# edit myapp.yaml to set the appAddress and optionally add components and/or parameters
```
Comment (jlewi) [inline kfctl/README.md]: Why are you introducing a myapp.yaml file here?

I thought the whole purpose of the CLI was to pretty much keep the existing semantics of kfctl.sh but just reuse the go library.

File: kfctl/README.md [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+
+  Calls KsService.Apply //RELEVANT
+```
+
+- Note: Where there is a RELEVENT comment indicates where we need to insert a shared Interface that kfctl can use.
+- Note: kApp refers to the ksonnet interface.
+
+## Proposed Design
+
+### 1. Move the common interface and types required for the library to a pkg directory under bootstrap
```
Comment (jlewi) [inline kfctl/README.md]: Should we define an interface for a Kubeflow App that would wrap ksonnet?

e.g

interface KsApp {
   generateK8s
  generateK8s
   applyK8s
   applyPlatform
}

This struct is what would be used by the CLI and webapp?

I guess an open question is how we would extend it to allow implementations for different platforms?

Comment (kkasravi) [reply kfctl/README.md]: > Should we define an interface for a Kubeflow App that would wrap ksonnet?
> 
> e.g
> 
> interface KsApp {
> generateK8s
> generateK8s
> applyK8s
> applyPlatform
> }
> 
> This struct is what would be used by the CLI and webapp?
> 
> I guess an open question is how we would extend it to allow implementations for different platforms?

Yes I'll define a struct or interface. I'll post here when the implementation is slightly farther along (a day or so hopefully)

Comment (jlewi) [reply kfctl/README.md]: SGTM. You could also just submit a skeleton PR with just the interface or struct but no implementation first.

**Comment by jlewi** [comment]
In terms of directory structure should we put the readme in the bootstrap directory and then plan on renaming the bootstrap to kfctl later on?

Or does it make sense to put the readme in the kfctl directory for now and just move the code later?

File: kfctl/README.md (hunk 5 above)
Comment (kkasravi) [reply kfctl/README.md]: +1 on getting the skeleton PR in early for review.

**Comment by kkasravi** [comment]
Maybe the former - bootstrap would look like:

bootstrap/cmd/kfctl
bootstrap/cmd/kfctl/README.md
bootstrap/cmd/kfctl/main.go
bootstrap/cmd/kfctl/cmd/{apply.go,generate.go, ...}
bootstrap/pkg/apis/apps/v1alpha1/kfapp_types.go
bootstrap/pkg/client/kfapi/typed/apps/v1alpha1/kfapp.go
bootstrap/pkg/server/main.go
bootstrap/pkg/server/app/{server.go,saKey.go, ...}
bootstrap/pkg/server/app/options/options.go

This is similar to how kubebuilder lays things out for CRDs (though it would have a controller subdir instead of server)

**Comment by jlewi** [comment]
That seems good to me.

nit: wouldn't server/main.go be under cmd? e.g. 

bootstrap/cmd/kfctlWeb/main.go

We should plan on

1. Renaming "bootstrap" to kfctl

   * I think the name bootstrap is confusing so we should get it rid of it when possible

1. Probably moving it into its own repo at some point
   
    * At which point maybe the toplevel "bootstrap" directory is unnecessary?

I'd like the code for the webserver to be in a separate directory so that we can use OWNERs files to narrowly restrict the approver list. The directory structure above supports this since all the webapp code would be in the server directory and we can configure OWNERs files to not inherit parents see [here](https://github.com/kubernetes/community/blob/master/contributors/guide/owners.md#owners). 

One of the things we need to ensure through code review is that there isn't any cross talk between requests; that could leak information. So we want to use a pattern like

```
func requestHandler() {
  app := NewKubeflowApp(...)
  defer app.Delete
  app.Generate()
}
```

This way its easy to check by inspection that we are being defensive in avoiding potential cross thread talking.


@kunmingg thoughts?

**Comment by kkasravi** [comment]
> nit: wouldn't server/main.go be under cmd? e.g.
> 
> bootstrap/cmd/kfctlWeb/main.go

yes, I think that makes more sense.

> I'd like the code for the webserver to be in a separate directory so that we can use OWNERs files to narrowly restrict the approver list. The directory structure above supports this since all the webapp code would be in the server directory and we can configure OWNERs files to not inherit parents see [here](https://github.com/kubernetes/community/blob/master/contributors/guide/owners.md#owners).

+1

**Comment by jlewi** [comment]
Thanks @kkasravi 
Should I LGTM this as is or there are additional changes you want to make first?

**Comment by kunmingg** [comment]
@jlewi 
Create https://github.com/kubeflow/kubeflow/issues/2152 to track container-level isolation per request.

**Comment by kkasravi** [comment]
@jlewi i think we're at a good point to close this PR. 
Additional issues can be opened that follow from this - like what @kunmingg is doing.

**Comment by jlewi** [comment]
@kkasravi By close do you mean LGTM and merge or close this without merging?  (I usually interpret close as don't merge). Why wouldn't we merge in the README seems useful to me?

**Pull request merged**